docs/specs/lcb-traits.md

```diff
@@ -0,0 +1,364 @@
+*and can be verified* at compile time.  This doesn't satisfy the requirement.
+
+For example, it would be convenient to write a `Log()` handler that can log
+any type that can also be formatted.  However, with the generic handlers scheme
+described above, `Log()` would have to take arguments as `any` and defer
+type-checking to runtime.
+
+## Traits
+
+Traits are collections of methods that can be used to extend the functionality
+of a type.
+
+Traits have no storage associated with them.  Storage is provided by
+implementing the trait for a specific storage type.
+
+### Defining traits
+
+    public trait type CanFormatAsString
+        method type FormatAsString(in me) as string
+    end type
+
+A trait definition is a block containing a list of zero or more `method type`
+declarations.  `method type` declarations take the same form as existing
+`handler type` declarations.
+
+Two additional symbols are defined within a trait definition:
+
+1. `MyType` is always equivalent to the type of the value that the method is
+   being called for.
+2. `me` must always be the first parameter a `method` definition or
+   `method type` declaration; it is equivalent to writing `me as MyType`.  The
+   `me` parameter must always be `in` or `inout`.
+
+### Implementing traits
+
+    public implementation of CanFormatAsString for T
+        method FormatAsString(in me) as string
+            variable tFormatted as string
+            -- Format me parameter
+            return tFormatted
+        end method
+    end implementation
+
+`method` definitions take the same form as existing `handler` definitions.
+
+In order for an `implementation of U for T` to be valid, the following criteria
+must be fulfilled:
+
+1. **Completeness Rule**.  All methods of `U` must be implemented (either in
+   the `implementation` block or via a default implementation in the trait
+   definition).
+2. **Locality Rule**.  Either `U` or `T` must be local, i.e. defined in the
+   current module.
+3. **Visibility Rule**.  The visibility of the implementation must be at least
+   the accessibility of `U` (i.e. if `U` is `public`, then the implementation
+   must also be `public`).
+
+#### Completeness rule
+
+The **Completeness Rule** is required to ensure that it is possible to call any
+method of a trait `U` for any value of type `T` for which `U` has been
+implemented.
+
+#### Locality rule
+
+The **Locality Rule** prevents two implementations of `T`.  Without this rule,
+the following scenario would be possible:
+
+* Module `A` defines and type `T`
+* Module `B` implements `CanFormatAsString` for `T`
+* Module `C` implements `CanFormatAsString` for `T` (differently)
+* Module `D` uses modules `B` and `C` and attempts to format an instance of `T`
+  as a string.  Which implementation should be used?
+
+By requiring either the trait type or the concrete type to be defined locally
+to the implementation, this diamond problem can never occur.
+
```

Comment:
@livecodestephen: I'm not sure this problem is resolved by derivation - as the derived type BobT will be distinct from the original AliceT. The result is that any modules which consume / produce AliceT will still be producing AliceT's - there would need to be explicit code in the all the modules wanting to get the CanFormatAsString effect on AliceT's to turn AliceT's into BobT's (recursively through any compound types too).

Stepping back slightly, the problem here comes down to the definition of 'world'.

Modules as we've currently been thinking about them have to interact with a world they do not know - this makes the locality rules required as it prevents (ahead of time) any potential conflicts arising in this most general context (which is what we have in the context of LCB-as-an-extension-language).

In the app example though, the person producing the app has complete control over the 'world'. The result of this is that only 'modules which might be used in an unknown world' have to conform to the locality rules - if you are compiling a module for a world you control then there is no problem... At the 'link' (or 'run' step) you might get an error if you've provided conflicting implementations of a trait - just as you would if you have an implementation of function twice in two object files and try to link them together.

Upshot: in becomes a property of the 'scope' of the module which determines whether you can define an implementation for a public trait and public type you do not define, and modules which don't have an appropriate scope cannot be used in a wider context (i.e. used by a third-party).

(Of course, there is a generalization here... It is not beyond the realms of possibility that I might want to provide a module which is intended for use by third parties but within whose 'world of operation' you want to adapt/specialise third-party types and traits but that's perhaps something for later consideration!)